onnx/defs/tensor/defs.cc

```diff
@@ -474,16 +477,38 @@ ONNX_OPERATOR_SET_SCHEMA(
           if (!hasNInputShapes(ctx, 1)) {
             return;
           }
-
+          auto input_type = ctx.getInputType(0);
+          if (!input_type->has_tensor_type()) {
+              throw InferenceError("Input 0 expected to have tensor type");
+          }
+          const TensorShapeProto& shape = input_type->tensor_type().shape();
           std::vector<int64_t> perm;
-          if (!getRepeatedAttribute(ctx, "perm", perm)) {
-            for (int i =
-                     ctx.getInputType(0)->tensor_type().shape().dim_size() - 1;
-                 i >= 0;
-                 --i) {
+          bool has_perm_attr = getRepeatedAttribute(ctx, "perm", perm);
+          if (!has_perm_attr) {
+            for (int i = shape.dim_size() - 1; i >= 0; --i) {
               perm.push_back(i);
             }
           }
+          else if(!perm.empty()){
+              //check if every index is valid
+              for (int64_t fromDimIndex : perm)
+                  if (!(0 <= fromDimIndex && fromDimIndex < shape.dim_size()) {
+                      std::ostringstream oss;
+                      oss << "Invalid attribute perm {" << perm[0];
+                      for (size_t i=1;i!=perm.size();++i) {
+                          oss << ", " << perm[i];
+                      }
+                      oss << "}, input shape = {";
+                      if (shape.dim_size() > 0) {
+                          oss<< shape.dim(0).dim_value();
+                          for (int i = 1; i != shape.dim_size(); ++i) {
+                              oss << ", " << shape.dim(i).dim_value();
```

Comment:
Why do you need to print the dim values? what user should know is `shape.dim_size()` right?

Reply:
It's not so obvious. Because they may be inferred from another node.  Currently, there is no way to view it.